### PR TITLE
Change AioCb to primarily use Bytes instead of Rc<[u8]>

### PR DESCRIPTION
**Pull request by asomers**
`Rc<[u8]>` isn't a very good buffer type to use for aio.  For one thing, it lacks interior mutability.  For another, a single `Rc<[u8]>` can't be carved up into smaller buffers of the same type.  `Bytes` and `BytesMut` fix both problems.  This PR removes the ability to construct an `AioCb` from `Rc<[u8]>` and adds the ability to construct one from `Bytes`, `BytesMut`, or raw pointers (for consumers who need even more flexibility).  At this stage, the PR has the following warts:

1. A hack is necessary to force small `Bytes` buffers to allocate on the heap.  I plan to fix this with an enhancement to the bytes crate.
2. The `AioCb::buffer` method is necessary due to a deficiency in the tokio-core crate.  Once I fix that, then only `AioCb::into_buffer`will need to be public.

### CODE REVIEW

File: src/sys/aio.rs [hunk 1]
```diff
@@ -86,10 +118,21 @@ pub struct AioCb<'a> {
...
-    keeper: Keeper<'a>
+    buffer: Buffer<'a>
 }
 
 impl<'a> AioCb<'a> {
+    /// Remove the inner `Buffer` and return it
+    ///
+    /// It is an error to call this method while the `AioCb` is still in
+    /// progress.
+    pub fn buffer(&mut self) -> Buffer<'a> {
```
Comment (Susurrus) [inline src/sys/aio.rs]: Does it make sense to make this return a `Result` type instead of having failing if the in progress invariant is broken?

**Comment by Susurrus** [comment]
Any movement on this?

**Comment by asomers** [comment]
Yeah, I'm 99% done modifying the downstream crates.  But I can't be confident that my new API will work until I'm 100% done.

**Comment by Susurrus** [comment]
Got an ETA on this? We're at 6 months since our last release, and I think it's smart that we get this out the door instead of waiting. How would you feel about releasing 0.10 now?

File: src/sys/aio.rs (hunk 1 above)
Comment (asomers) [reply src/sys/aio.rs]: I think not.  If we do that, then `AioCb::into_buffer` will need to be able to handle it, such as by returning `None`.  However, that method can already return `Buffer::none`, which means something different.  It would be confusing to return two different None types.

In my experience, your program's flow should make it obvious whether any given `AioCb` is still in progress or not.  Treating an in-progress `AioCb` as though it has already finished is almost always going to be a programming bug.  So I think `assert`ing is appropriate.

**Comment by asomers** [comment]
I'm all satisfied with this change now.  Review at your leisure.

File: src/sys/aio.rs [hunk 2]
```diff
@@ -66,17 +67,48 @@ pub enum AioCancelStat {
...
+/// the kernel has a pointer to it.
 #[derive(Clone, Debug)]
-enum Keeper<'a> {
+pub enum Buffer<'a> {
     none,
-    /// Keeps a reference to a Boxed slice
-    boxed(Rc<Box<[u8]>>),
+    /// Immutable shared ownership `Bytes` object
+    // Must use out-of-line allocation so the address of the data will be
+    // stable.  Bytes and BytesMut sometimes dynamically allocate a buffer, and
```
Comment (Susurrus) [inline src/sys/aio.rs]: Backticks around Bytes and BytesMut

File: src/sys/aio.rs [hunk 3]
```diff
@@ -66,17 +67,48 @@ pub enum AioCancelStat {
...
 }
 
-/// Private type used by nix to keep buffers from Drop'ing while the kernel has
-/// a pointer to them.
+/// Owns (uniquely or shared) a memory buffer to keep it from `Drop`ing while
+/// the kernel has a pointer to it.
 #[derive(Clone, Debug)]
-enum Keeper<'a> {
+pub enum Buffer<'a> {
     none,
```
Comment (Susurrus) [inline src/sys/aio.rs]: Can you add a doccomment for this variant?

Additionally, why is this lowercase `none`? I'd expect it to be `None`.

File: CHANGELOG.md [hunk 4]
```diff
@@ -107,6 +105,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
   ([#648](https://github.com/nix-rust/nix/pull/648))
 
 ### Removed
+- `AioCb::from_boxed_slice` has been removed.  It was never actually safe.
```
Comment (Susurrus) [inline CHANGELOG.md]: What would be the replacement for this kind of functionality? If there's a short answer, it should be said here, otherwise it should be stated to "read the docs for more details" or "see the PR for more details". We should make it as easy as possible for downstream users of our library to update their code, this means pointing them in the right direction when we break it.

File: Cargo.toml [hunk 5]
```diff
@@ -18,6 +18,11 @@ bitflags = "1.0"
 cfg-if = "0.1.0"
 void = "1.0.2"
 
+# Don't include the optional serde feature
```
Comment (Susurrus) [inline Cargo.toml]: Can you move this comment to right about the `default-features` line so it's more clear that's what it's pertaining to?

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,7 +7,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Added
 - Added `AioCb::from_ptr` and `AioCb::from_mut_ptr`
-  ([#TODO](https://github.com/nix-rust/nix/pull/TODO))
+  ([#820](https://github.com/nix-rust/nix/pull/820))
```
Comment (Susurrus) [inline CHANGELOG.md]: Please fix the initial commit to have this number from the start.

File: CHANGELOG.md [hunk 7]
```diff
@@ -58,8 +58,6 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ### Changed
 - Use native `pipe2` on all BSD targets.  Users should notice no difference.
   ([#777](https://github.com/nix-rust/nix/pull/777))
-- `AioCb::from_boxed_slice` no longer allows `aio_read`
```
Comment (Susurrus) [inline CHANGELOG.md]: Is there a reason why we have this step fixing up `from_boxed_slice` only to remove it in the following commit?

Comment (asomers) [reply CHANGELOG.md]: Not really.  I should squash them together.

File: src/sys/aio.rs (hunk 3 above)
Comment (asomers) [reply src/sys/aio.rs]: Not for any good reason.  I'll change it back.  However, I'm wondering if I should make the Buffer type private.  If I do that, then I'll have to replace `AioCb::into_buffer` with `AioCb::into_bytes -> Option<Bytes>` and `AioCb::into_bytes_mut -> Option<BytesMut>`.  Do you think that would be a better API?

Comment (asomers) [reply src/sys/aio.rs]: Nevermind.  I forgot that I can't do that due to a limitation of tokio-core.  In order to satisfy tokio-core, I need way to get the `Buffer` without consuming the `AioCb`.

**Comment by asomers** [comment]
@Susurrus I think I responded to all of your concerns except the one about squashing.  I left everything unsquashed for your viewing pleasure.

File: src/sys/aio.rs [hunk 8]
```diff
@@ -1,4 +1,27 @@
...
+//! [`fsync`](struct.AioCb.html#method.fsync) operations.  Completion
+//! notifications can optionally be delivered via
+//! [signals](../signal/enum.SigevNotify.html#variant.SigevSignal), via the
+//! [`aio_suspend`](fn.aio_suspend.html) function, or via polling.  Some
+//! platforms support other completion
+//! notifications, such as
+//! [kevent](../signal/enum.SigevNotify.html#variant.SigevKevent).
+//!
+//! Multiple operations may be submitted in a batch with
+//! [`lio_listio`](fn.lio_listio.html), thought the standard does not guarantee
```
Comment (Susurrus) [inline src/sys/aio.rs]: thought -> though

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,9 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 
 ### Added
+- Added `AioCb::from_ptr` and `AioCb::from_mut_ptr`
+  ([#820](https://github.com/nix-rust/nix/pull/TODO))
```
Comment (Susurrus) [inline CHANGELOG.md]: This first link should be removed.

File: src/sys/aio.rs [hunk 10]
```diff
@@ -231,6 +617,46 @@ impl<'a> AioCb<'a> {
...
 
     /// Cancels an outstanding AIO request.
+    ///
+    /// The operating system is not required to implement cancellation for all
+    /// file and device types.  Even if it does, there is no guarantee that the
+    /// operation has not already completed.  So the caller must check the
+    /// result and handle operations that were not canceled or that have already
+    /// completed.
+    ///
+    /// See also
```
Comment (Susurrus) [inline src/sys/aio.rs]: I would rephrase to indicate this is an external link like "See the man pages for `aio_cancel` for more details." This applies for all the uses of "See also LINK".

**Comment by asomers** [comment]
Travis is back up now.  I'm going to go ahead and merge this even though the last OSX builds never ran, because the last changes I made were just formatting stuff that shouldn't have broken OSX.

bors r+ susurrus

**Comment by Susurrus** [comment]
🤞

**Comment by asomers** [comment]
Try try again

bors r+ susurrus

**Comment by asomers** [comment]
If at third you don't succeed, try try again.

bors r+ susurrus

**Comment by Susurrus** [comment]
Travis is still having a tough time with Mac builds. Let's wait a couple of days until the backlog dies down a bit and we can try again.

**Comment by asomers** [comment]
I see that libc was able to run OSX builds on Travis.  Maybe we can too.

bors r+ susurrus

**Comment by Susurrus** [comment]
`libc` has a paid Travis subscription, so they can almost always run jobs. Note that the Open Source Mac build queue is almost 3000; this isn't going to finish.

**Comment by Susurrus** [comment]
Mac OS build queue down to 1100, so maybe this could actually happen now!

bors r+
